zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -93,35 +109,100 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
     // - Index of the "Root" field elements: {0}
     // - Index of the "NullifierS" field elements: [1, NumInputs + 1[
     // - Index of the "CommitmentS" field elements: [NumInputs + 1, NumOutputs + NumInputs + 1[
-    // - Index of the "v_pub_in" field element: {NumOutputs + NumInputs + 1}
-    // - Index of the "v_pub_out" field element: {NumOutputs + NumInputs + 1 + 1}
+    // - Index of the "h_sig" field element: {NumOutputs + NumInputs + 1}
+    // - Index of the "Message Authentication TagS" (h_i) field elements: [NumOutputs + NumInputs + 1 + 1, NumOutputs + NumInputs + 1 + NumOuputs [
+    // - Index of the "Residual Field Element(s)" field elements: [NumOutputs + NumInputs + 1 + NumOuputs + 1 , NumOutputs + NumInputs + 1 + NumOuputs + 1 + nb_field_residual]
+    //
+    // The Residual field elements are structured as follows:
+    // - v_pub_in [0, size_value[
+    // - v_pub_out [size_value + 1, 2*size_value[
+    // - h_sig remaining bits [2*size_value + 1, 2*size_value + (digest_length-field_capacity)[
+    // - nullifierS remaining bits [2*size_value + (digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs)*(digest_length-field_capacity)[
+    // - commitmentS remaining bits [2*size_value + (1+NumInputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity)[
+    // - message authentication tagS remaining bits [2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+2*NumInputs+NumOutputs)*(digest_length-field_capacity)]
     // ============================================================================================ //
 
+    // This function extract the remaining bits of the nullifierS, commitmentS or h_iS from the residual field element(S)
+    function extract_extra_bits(uint start, uint end, uint[] memory primary_inputs) public pure returns (bytes1) {
+        // We assume we extract less than 2 bytes
+        require(
+            end-start < 16,
+            "invalid range"
+        );
+        uint16 res;
+
+        // We first compute the offset (to forego the v_pubs and h_sig remaining bits)
+        uint offset = 2 * size_value + (digest_length - field_capacity);
+
+        // The remaining bits could be written over two field elements
+        // as such we compute the indices of the first and last field elements
+        // they are located at
+        uint index_start_bytes32 = 1 + 1 + 2*jsIn + jsOut + (offset+start)/field_capacity;
+        uint index_end_bytes32 = 1 + 1 + 2*jsIn + jsOut + (offset+end)/field_capacity;
+
+        // We compute the padding length in the field elements containing the remaining bits
+        uint padding_start = digest_length - field_capacity;
+        uint padding_end = padding_start;
+        if (primary_inputs.length-1 == index_end_bytes32) {
+            padding_end = digest_length - ( (offset + (digest_length - field_capacity)*(2*jsIn + jsOut)) % field_capacity);
```

Comment:
You `padding` variables are confusing. You subtract the offset, then add it again in `uint index_start_byte = (padding_start+offset+start)/8;`

Please simplify things that can be, or at very least add comments to document your code.

Reply:
Yes. I will add some comments.

The offset represents the 128 bits of the public values and 3 bits for the h_sig, while the padding represent just padding. What I am trying to do here is say that the padding is 3 bit if the field element is full, else 256 - the remaining bits.

Reply:
Corrected